File: backend/src/main/java/com/g2u/admin/domain/product/Product.java
package com.g2u.admin.domain.product;

import com.g2u.admin.domain.category.Category;
import com.g2u.admin.domain.common.TenantAwareEntity;
import com.g2u.admin.domain.tag.Tag;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.BatchSize;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Entity
@Table(name = "products")
@SQLDelete(sql = "UPDATE products SET deleted_at = NOW() WHERE id = ? AND version = ?")
@SQLRestriction("deleted_at IS NULL")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Product extends TenantAwareEntity {

    @Column(nullable = false, length = 300)
    private String name;

    @Column(nullable = false, length = 300)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "short_description", length = 500)
    private String shortDescription;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private ProductStatus status = ProductStatus.DRAFT;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private ProductSource source = ProductSource.OWN;

    @Column(precision = 12, scale = 2)
    private BigDecimal price;

    @Column(name = "compare_at_price", precision = 12, scale = 2)
    private BigDecimal compareAtPrice;

    @Column(length = 3)
    @Builder.Default
    private String currency = "UAH";

    @Column(length = 100)
    private String sku;

    @Column(length = 100)
    private String barcode;

    @Column(name = "track_inventory")
    @Builder.Default
    private boolean trackInventory = false;

    @Builder.Default
    private Integer quantity = 0;

    @Column(name = "low_stock_threshold")
    private Integer lowStockThreshold;

    @Column(precision = 8, scale = 2)
    private BigDecimal weight;

    @Column(name = "weight_unit", length = 5)
    @Builder.Default
    private String weightUnit = "kg";

    @Column(name = "meta_title", length = 200)
    private String metaTitle;

    @Column(name = "meta_description", length = 500)
    private String metaDescription;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "JSON")
    private Map<String, Object> attributes;

    @Column(name = "published_at")
    private Instant publishedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @OrderBy("sortOrder ASC")
    @BatchSize(size = 20)
    @Builder.Default
    private List<ProductVariant> variants = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @OrderBy("displayOrder ASC")
    @BatchSize(size = 20)
    @Builder.Default
    private List<ProductOption> options = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @OrderBy("sortOrder ASC")
    @BatchSize(size = 20)
    @Builder.Default
    private List<ProductMedia> media = new ArrayList<>();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "product_categories",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    @BatchSize(size = 20)
    @Builder.Default
    private Set<Category> categories = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "product_tags",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @BatchSize(size = 20)
    @Builder.Default
    private Set<Tag> tags = new HashSet<>();
}

File: frontend/src/lib/types/product.ts
export interface Product {
	id: string;
	name: string;
	slug: string;
	description: string;
	shortDescription: string;
	status: 'DRAFT' | 'ACTIVE' | 'ARCHIVED';
	price: number;
	compareAtPrice: number | null;
	currency: string;
	sku: string;
	barcode: string;
	trackInventory: boolean;
	quantity: number;
	weight: number | null;
	weightUnit: string;
	metaTitle: string;
	metaDescription: string;
	attributes: Record<string, unknown>;
	publishedAt: string | null;
	createdAt: string;
	updatedAt: string;
	variants: ProductVariant[];
	media: ProductMedia[];
	categories: CategoryRef[];
	tags: TagRef[];
}

export interface ProductVariant {
	id: string;
	name: string;
	sku: string;
	price: number;
	compareAtPrice: number | null;
	costPrice: number | null;
	quantity: number;
	lowStockThreshold: number | null;
	options: Record<string, unknown>;
	sortOrder: number;
	active: boolean;
}

export interface ProductMedia {
	id: string;
	url: string;
	altText: string;
	mediaType: string;
	sortOrder: number;
	primary: boolean;
}

export interface CategoryRef {
	id: string;
	name: string;
	slug: string;
}

export interface TagRef {
	id: string;
	name: string;
	slug: string;
}

/** Lightweight product type returned by the list endpoint (no collections). */
export interface ProductListItem {
	id: string;
	name: string;
	slug: string;
	status: 'DRAFT' | 'ACTIVE' | 'ARCHIVED';
	price: number;
	compareAtPrice: number | null;
	currency: string;
	sku: string;
	trackInventory: boolean;
	quantity: number;
	publishedAt: string | null;
	createdAt: string;
	updatedAt: string;
	primaryImageUrl: string | null;
	source: 'OWN' | 'MARKETPLACE';
}

export interface Page<T> {
	content: T[];
	totalElements: number;
	totalPages: number;
	size: number;
	number: number;
	first: boolean;
	last: boolean;
}

/** Option value (e.g. "Red", "XL") */
export interface OptionValue {
	id: string;
	value: string;
	displayOrder: number;
}

/** Product option (e.g. "Color", "Size") with its values */
export interface ProductOption {
	id: string;
	name: string;
	displayOrder: number;
	values: OptionValue[];
}

/** A specific variant of a product with its own SKU/price/availability */
export interface ProductVariantDetail {
	id: string;
	sku: string;
	price: number;
	compareAtPrice: number | null;
	available: boolean;
	optionValues: OptionValue[];
	createdAt: string;
	updatedAt: string;
}
